Sources/NIO/EventLoop.swift

```diff
@@ -273,13 +273,47 @@ public protocol EventLoop: EventLoopGroup {
     /// Contrary to `makeSucceededFuture`, `makeSucceededVoidFuture` is a customization point for `EventLoop`s which
     /// allows `EventLoop`s to cache a pre-succeded `Void` future to prevent superfluous allocations.
     func makeSucceededVoidFuture() -> EventLoopFuture<Void>
+
+    /// Debug hook: track a promise creation and its location.
+    ///
+    /// This debug hook is called by EventLoopFutures and EventLoopPromises when they are created, and tracks the location
+    /// of their creation. It combines with `_promiseCompleted` to provide high-fidelity diagnostics for debugging leaked
+    /// promises.
+    ///
+    /// In release mode, this function will never be called.
+    ///
+    /// It is valid for an `EventLoop` not to implement any of the two `_promise` functions. If either of them are implemented,
+    /// however, both of them should be implemented.
+    func _promiseCreated(futureIdentifier: _NIOEventLoopFutureIdentifier, file: StaticString, line: UInt)
```

Comment:
Maybe a silly question, but why does this need to be framed in terms of a future identifier rather than just a future and having the `EventLoop` implementation deal with the identity of that future?

Reply:
The short reason is that I wanted to persist the requirement that led to `SelectableEventLoop` using the somewhat odd `obfuscatePointerValue` function. If we pass the future to the callback directly, users will likely naively store this in a dictionary using `ObjectIdentifier`. This will have the effect of causing the Xcode memory graph debugger to think that the future is being retained by the dictionary. It won't be: object identifiers don't retain objects, but the memory graph debugger doesn't know that.

This way we can arrange for an identifier that is stably defined in terms of the object identifier but is not confusable for a pointer to a given object. While we're here, it also discourages implementers from being "clever" and using this to observe promise creation and destruction for any reason other than this one.